NitroxPatcher/Patches/Dynamic/PickPrefab_AddToContainerAsync_Patch.cs

```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.Reflection;
+using System.Reflection.Emit;
+using HarmonyLib;
+using NitroxClient.Communication.Abstract;
+using NitroxModel.DataStructures;
+using NitroxModel.Helper;
+using NitroxModel.Packets;
+using UnityEngine;
+
+namespace NitroxPatcher.Patches.Dynamic;
+
+/// <summary>
+/// Sync destruction of giver object when required.
+/// </summary>
+public sealed partial class PickPrefab_AddToContainerAsync_Patch : NitroxPatch, IDynamicPatch
+{
+    internal static readonly MethodInfo TARGET_METHOD = AccessTools.EnumeratorMoveNext(Reflect.Method((PickPrefab t) => t.AddToContainerAsync(default, default)));
+
+    /*
+     * 1st injection:
+     * if (!component)
+     * {
+     *     UnityEngine.Object.Destroy(gameObject);
+     *     BroadcastDeletion(this);      <------- [INSERTED LINE]
+     *  
+     *  2nd injection:
+     *  else if (!container.HasRoomFor(component))
+     *  {
+     *     UnityEngine.Object.Destroy(gameObject);
+     *     BroadcastDeletion(this);      <------- [INSERTED LINE]
+     */
+    public static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
+    {
+        // The 2 injections are similar (looking for a destroy instruction and adding our callback after it)
+        return new CodeMatcher(instructions).MatchEndForward(new CodeMatch(OpCodes.Call, Reflect.Method(() => Object.Destroy(default))))
+                                            .Advance(1)
+                                            .InsertAndAdvance([
+                                                new CodeInstruction(OpCodes.Ldarg_0),
+                                                new CodeInstruction(OpCodes.Call, Reflect.Method(() => BroadcastDeletion(default)))
+                                            ])
+                                            .MatchEndForward(new CodeMatch(OpCodes.Call, Reflect.Method(() => Object.Destroy(default))))
+                                            .Advance(1)
+                                            .InsertAndAdvance([
+                                                new CodeInstruction(OpCodes.Ldarg_0),
+                                                new CodeInstruction(OpCodes.Call, Reflect.Method(() => BroadcastDeletion(default)))
+                                            ])
+                                            .InstructionEnumeration();
```

Comment:
CodeMatcher has support for repeated operations, don't know if it's better here:
```
new CodeMatcher(instructions)
    .MatchForward(false, // false = move at the start of the match, true = move at the end of the match
        new CodeMatch(OpCodes.Stfld),
        new CodeMatch(OpCodes.Ldarg_0),
        new CodeMatch(i => i.opcode == OpCodes.Ldfld && ((FieldInfo)i.operand).Name == "test"))
    .Repeat( matcher => // Do the following for each match
                 matcher
                   .Advance(2) // Move cursor to before ldfld
                   .InsertAndAdvance(
                      new CodeInstruction(OpCodes.Dup),
                      new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(Foo), "Foo"))
                   )
    )
    .InstructionEnumerable(); // Finally, return the manipulated method instructions
```

Reply:
I'll do it this time, but we should probably refrain from doing it everywhere because we could lose context in case Subnautica code gets updated. Even though tests are there to show us that some patch broke (e.g. now there are 3 matches and not 2), we might lose the exact intention of which matches exactly we wanted to patch, and which one we didn't want to patch.